### PR TITLE
"My" search

### PR DESCRIPTION
**Pull request by mrvisser**
This should support the "My Stuff".

There still needs to be some UI stuff associated to this, which is in development. So it's possible this doesn't turn out to meet 100% of the UI requirements. However, here it is for initial review if desired.

Built on top of GroupProfiles, of course. So that will need to be merged first.

### CODE REVIEW

**Comment by mrvisser** [comment]
Rebased with master, self-reviewed. Assigning to @simong for review.

File: node_modules/oae-util/tests/test-globals.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+describe('Globals', function() {
+
+    describe('Underscore', function() {
+
+        describe('oaeObj', function() {
+
+            /**
+             * Test that verifies that an object is created from a variety of input
+             */
+            it('verify an object is created from a variety of input', function() {
```
Comment (simong) [inline node_modules/oae-util/tests/test-globals.js]: I personally am fine with not adding `callback` in the arguments and not doing `return callback();` for synchronous tests, but I believe we settled on doing so

File: node_modules/oae-authz/tests/test-groups.js [hunk 2]
```diff
@@ -15,10 +15,12 @@
 
 var _ = require('underscore');
 var assert = require('assert');
+var shortid = require('shortid');
```
Comment (simong) [inline node_modules/oae-authz/tests/test-groups.js]: `ShortId`

**Comment by simong** [comment]
At this point, I think we need a place to communicate to REST clients what `scope`s, `resourceType`s, etc.. are. Once https://github.com/oaeproject/Hilary/issues/1148 gets addressed we will be able to expose a pre-amble that goes into more detail about each endpoint. I think it'd be useful to document each search "type" that we register through `registerSearch` with Swagger separately. ie, have separate docs for:
 * /api/search/general
 * /api/search/deleted
 * /api/search/following
 * /api/search/followers
 * /api/search/{library-name}

WDYT?

**Comment by mrvisser** [comment]
Yea, I agree. It would be great to be able "register" your own docs for it. Maybe it's actually better for now if we don't have the central doc, but instead have individual doc blocks for each search type that gets registered?

**Comment by mrvisser** [comment]
Hopefully swagger has a way to document an enumerable string, that way we can have a semantic way of documenting the different resource types, scopes, etc...

**Comment by simong** [comment]
Only had some superminor feedback, back to @mrvisser

File: node_modules/oae-authz/tests/test-groups.js (hunk 2 above)
Comment (mrvisser) [inline node_modules/oae-authz/tests/test-groups.js]: Why does this one get camel cased while all the other external deps are lowercased? I'd suggest just renaming the others to `shortid` so the style is more predictable

**Comment by mrvisser** [comment]
Assigning to @simong for feedback on `shortid` dependency

File: node_modules/oae-util/tests/test-globals.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                    '0': 0,
+                    'null': null,
+                    'anotherNull': null,
+                    'another0': 0
+                };
+
+                // Ensure the result is the expected object, and that `source` gets overridden
+                assert.deepEqual(_.oaeExtendDefined(source, extendWith), expectedDest);
+                assert.deepEqual(source, expectedDest);
+                callback();
```
Comment (simong) [inline node_modules/oae-util/tests/test-globals.js]: You'll have to put `callback` in `it`'s parameter list

Comment (simong) [inline node_modules/oae-util/tests/test-globals.js]: You'll also have to `return` them

Comment (mrvisser) [inline node_modules/oae-util/tests/test-globals.js]: > You'll also have to `return` them

I thought we didn't `return` them for "simple methods" that didn't have async calls / error handling cases ?

File: node_modules/oae-authz/tests/test-groups.js (hunk 2 above)
Comment (simong) [inline node_modules/oae-authz/tests/test-groups.js]: I'm not sure. I think we started using PascalCasing all variables requiring modules of our own (e.g.,  PrincipalsDAO, ContentAPI, ..) but kept short single word variables such as `fs`, `assert`, `util`, ... lowercase? Either way it's all terribly inconsistent. I'd rather we keep using `ShortId` everywhere for now or settle on style and start changing all of them.

Comment (mrvisser) [inline node_modules/oae-authz/tests/test-groups.js]: > I'm not sure. I think we started using PascalCasing all variables requiring modules of our own (e.g., PrincipalsDAO, ContentAPI, ..) but kept short single word variables such as fs, assert, util, ... lowercase?

I think this is what we did as well. Can we settle on it? I'll rename all `ShortId`s to `shortid`, it's not too much.

**Comment by mrvisser** [comment]
I've defined the callback. Oops. Also rebased w/ master. Back to @simong .

Note this should simply go on hold as the UI portion has not gone under any review yet and it is fairly significant.

**Comment by simong** [comment]
LGTM, marking as "On hold"

**Comment by nicolaasmatthijs** [comment]
@simong : If this is still ready for merging, feel free to pull the trigger

**Comment by simong** [comment]
@nicolaasmatthijs This was On Hold pending the UI work. I'm happy to merge it in once the UI's ready

**Pull request merged**